### PR TITLE
[DELETE with DVs] Allow reading CDF from files with Deletion Vectors - Part 1/2

### PR DESCRIPTION
**Pull request by xupefei**
## Description

This PR is part of https://github.com/delta-io/delta/issues/1701. A detailed overview of changes is described at https://github.com/delta-io/delta/issues/1701.

This is the first PR to add support to allow reading CDC from files that have DV associated. In this PR we do some preparation work to allow fine control of how to handle masked rows: keep or drop. Later these two types will be used by CDCReader to pull masked rows out from files.

## How was this patch tested?

Existing test because it's refactor-only.

### CODE REVIEW

**Comment by xupefei** [comment]
@vkorukanti Could you review this PR?

File: core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala [hunk 1]
```diff
@@ -171,16 +169,24 @@ object ScanWithDeletionVectors {
...
-      tahoeFileIndex.matchingFiles(Seq(TrueLiteral), Seq(TrueLiteral))
-        .filter(_.deletionVector != null)
-    val filePathToDVMap = filesWithDVs
-      .map(x =>
-         absolutePath(tahoeFileIndex.path.toString, x.path).toUri -> x.deletionVector)
-      .toMap
+    val filesWithDVs = tahoeFileIndex
+      .matchingFiles(partitionFilters = Seq(TrueLiteral), dataFilters = Seq(TrueLiteral))
+      .filter(_.deletionVector != null)
+    val filePathToDVMap = filesWithDVs.map { addFile =>
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:181]: Can you add some comments here when the filter type is not present and why we need to default to IF_CONTAINED?

File: core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala [hunk 2]
```diff
@@ -171,16 +169,24 @@ object ScanWithDeletionVectors {
...
+      val key = absolutePath(tahoeFileIndex.path.toString, addFile.path).toUri
+      val filterType =
+        filterTypes.getOrElse(addFile.path, RowIndexFilterType.IF_CONTAINED)
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:184]: is the look by `String` works? I remember we need to use the `.toUri.toString` in order to have the unique and complete path.

@larsk-db please check if this is ok.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala [hunk 3]
```diff
@@ -74,7 +76,8 @@ class DeltaParquetFileFormatSuite extends QueryTest
 
         val fs = addFilePath.getFileSystem(hadoopConf)
         val broadcastDvMap = spark.sparkContext.broadcast(
-          Map(fs.getFileStatus(addFilePath).getPath().toUri -> dv)
+          Map(fs.getFileStatus(addFilePath).getPath().toUri ->
+            DeletionVectorDescriptorWithFilterType(dv, RowIndexFilterType.IF_CONTAINED))
```
Comment (vkorukanti) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala:80]: Can we extend this to test to test IF_NOT_CONTAINED filter as well? May be one more test combination.

**Review by vkorukanti — COMMENTED** [review]
LGTM.

few comments and clarification questions. Also the description in the PR doesn't point to the correct ticket with the design details. If there is none exists, could you please create an issue with design description on how reading CDF from Delta tables with DVs is supported.

File: core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala (hunk 2 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala:184]: We do lookups using `URI` in `DeltaParquetFileFormat`. IMO it's more robust than translating them into strings.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala (hunk 3 above)
Comment (xupefei) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaParquetFileFormatSuite.scala:80]: Done!

**Review by vkorukanti — APPROVED** [review]
lgtm

**Pull request closed without merging**